### PR TITLE
Adding BlocklyDevTools.Analytics

### PR DESCRIPTION
**Pull request by AnmAtAnm**
Adding BlocklyDevTools.Analytics, an interface for integrating an analytics library to track basic usage, including:
 * navigation.
 * saving, importing, and exporting.
 * warnings and errors.

---
This change is [Reviewable](https://reviewable.io/reviews/google/blockly/1217)

### CODE REVIEW

File: demos/blockfactory/analytics.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Stubbed interface functions for analytics integration.
```
Comment (rachel-fenichel) [inline demos/blockfactory/analytics.js]: Missing @author

File: demos/blockfactory/analytics.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ */
+
+goog.provide('BlocklyDevTools.Analytics');
+
+BlocklyDevTools.Analytics.LOG_TO_CONSOLE = false;
+
+/**
+ * Initializes the analytics framework, including noting that the page/app was
```
Comment (rachel-fenichel) [inline demos/blockfactory/analytics.js]: Missing @private/@package/@public for everything.

File: demos/blockfactory/analytics.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ */
+BlocklyDevTools.Analytics.onImport = function(typeId) {
+  // stub
+  this.LOG_TO_CONSOLE && console.log('Analytics.onImport(' + typeId + ')');
+};
+
+/**
+ * Event noting a project resource was saved. In the web Block Factory, this
+ * means downloaded to the user's system.
+ * @param typeId {string} An identifying string for the exported type.
```
Comment (rachel-fenichel) [inline demos/blockfactory/analytics.js]: Missing doc for optFormalId param.

File: demos/blockfactory/app_controller.js [hunk 4]
```diff
@@ -137,9 +140,12 @@ AppController.prototype.exportBlockLibraryToFile = function() {
   // Download file if all necessary parameters are provided.
   if (filename) {
     FactoryUtils.createAndDownloadFile(blockLibText, filename, 'xml');
+    BlocklyDevTools.Analytics.onExport('BlockFactoryLibrary.xml');
```
Comment (rachel-fenichel) [inline demos/blockfactory/app_controller.js]: The strings that indicate the type should be constants defined in BlocklyDevTools.Analytics.  That can be done in a followup as you flesh this out.

**Comment by AnmAtAnm** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[demos/blockfactory/analytics.js, line 22 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-22:-Komhnb8_C54st6DzR_0:b-gfm10b) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L22)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing @author
</blockquote></details>

Avoiding duplication and any eventual inaccuracy.  One can always look to git blame.

---

*[demos/blockfactory/analytics.js, line 30 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-30:-KomhoyzaQQlV8rScheG:b-896fix) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L30)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing @private/@package/@public for everything.
</blockquote></details>

Done.

---

*[demos/blockfactory/analytics.js, line 69 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl7-r1-69:-Komi3Bj6Nsmaquhm9q3:b-k1w1zb) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/analytics.js#L69)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

Missing doc for optFormalId param.
</blockquote></details>

Hmm... Forgot I even added this.  Replaced it with optMetadata to also store platform info.
Added JSDoc.

---

*[demos/blockfactory/app_controller.js, line 143 at r1](https://reviewable.io:443/reviews/google/blockly/1217#-Koic8LXJLO19FEVbUl8-r1-143:-KompOEK74otVSfUwNhu:b-896fix) ([raw file](https://github.com/google/blockly/blob/c15aee18b53e254934ab9f8ada2fe6daa2a8b44e/demos/blockfactory/app_controller.js#L143)):*
<details><summary><i>Previously, rachel-fenichel (Rachel Fenichel) wrote…</i></summary><blockquote>

The strings that indicate the type should be constants defined in BlocklyDevTools.Analytics.  That can be done in a followup as you flesh this out.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217)*

**Comment by rachel-fenichel** [comment]
Missing @const {string} on all of your constants.

I expected your functions to be package, not public.

---

Reviewed 2 of 6 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-KomvTeOeGUQIBs-XMpx:biwvy5l)*

**Comment by AnmAtAnm** [comment]
They need to be accessible from `workspacefactory` directory, at least for `wfactory_controller`.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-KonSAgFZ6ghzDVKSuEE:b-451q8m)*

**Comment by rachel-fenichel** [comment]
:lgtm:

---

Reviewed 1 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/blockly/1217#-:-Konmxmv31OFkVYrm9bJ:bnfp4nl)*

**Pull request merged**